File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/circuitbreakers/CircuitBreakerService.java
package uk.nhs.nhsx.circuitbreakers;

import uk.nhs.nhsx.circuitbreakers.utils.TokenGenerator;
import uk.nhs.nhsx.core.Jackson;
import uk.nhs.nhsx.core.aws.ssm.Parameter;

import java.util.Optional;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CircuitBreakerService {

    private static final Pattern CIRCUIT_BREAKER_RESOLUTION_PATH_PATTERN = Pattern.compile("/circuit-breaker/[\\-\\w]+/resolution/(?<pollingToken>.*)");

    private final Parameter<ApprovalStatus> initial;
    private final Parameter<ApprovalStatus> poll;

    public CircuitBreakerService(Parameter<ApprovalStatus> initial, Parameter<ApprovalStatus> poll) {
        this.initial = initial;
        this.poll = poll;
    }

    public CircuitBreakerResult getApprovalToken() {
        String token = TokenGenerator.getToken();

        TokenResponse tokenResponse = new TokenResponse();
        tokenResponse.setApprovalToken(token);
        tokenResponse.setApproval(initial.value().getName());

        return CircuitBreakerResult.ok(Jackson.toJson(tokenResponse));
    }

    public CircuitBreakerResult getResolution(String path) {
        Optional<String> pollingToken = extractPollingToken(path);

        if (pollingToken.isEmpty()) {
            return CircuitBreakerResult.missingPollingTokenError();
        }

        ResolutionResponse resolutionResponse = new ResolutionResponse();
        ApprovalStatus status = poll.value();
        resolutionResponse.setApproval(status.getName());

        return CircuitBreakerResult.ok(Jackson.toJson(resolutionResponse));
    }

    public static Optional<String> extractPollingToken(String path) {
        return Optional.ofNullable(path).flatMap(it -> {
            Matcher matcher = CIRCUIT_BREAKER_RESOLUTION_PATH_PATTERN.matcher(it);
            if (matcher.matches()) {
                String pollingToken = matcher.group("pollingToken");
                if (pollingToken != null && !pollingToken.isEmpty()) return Optional.of(pollingToken);
            }
            return Optional.empty();
        });
    }

    public static Predicate<String> startsWith(String path) {
        return (candidate) -> candidate.startsWith(path);
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/circuitbreakers/CircuitBreakerService.kt
package uk.nhs.nhsx.circuitbreakers

import uk.nhs.nhsx.circuitbreakers.CircuitBreakerResult.Companion.missingPollingTokenError
import uk.nhs.nhsx.circuitbreakers.CircuitBreakerResult.Companion.ok
import uk.nhs.nhsx.circuitbreakers.utils.TokenGenerator.token
import uk.nhs.nhsx.core.Json.toJson
import uk.nhs.nhsx.core.aws.ssm.Parameter
import java.util.Optional
import java.util.function.Predicate
import java.util.regex.Pattern

class CircuitBreakerService(
    private val initial: Parameter<ApprovalStatus>,
    private val poll: Parameter<ApprovalStatus>
) {
    fun getApprovalToken() = ok(toJson(TokenResponse(token, initial.value().statusName)))

    fun getResolution(path: String?) = when {
        extractPollingToken(path).isEmpty -> missingPollingTokenError()
        else -> ok(toJson(ResolutionResponse(poll.value().statusName)))
    }

    companion object {
        private val CIRCUIT_BREAKER_RESOLUTION_PATH_PATTERN =
            Pattern.compile("/circuit-breaker/[\\-\\w]+/resolution/(?<pollingToken>.*)")

        fun extractPollingToken(path: String?): Optional<String> = Optional.ofNullable(path).flatMap {
            val matcher = CIRCUIT_BREAKER_RESOLUTION_PATH_PATTERN.matcher(it)
            if (matcher.matches()) {
                val pollingToken = matcher.group("pollingToken")
                when {
                    pollingToken != null && pollingToken.isNotEmpty() -> Optional.of(pollingToken)
                    else -> Optional.empty()
                }
            } else Optional.empty()
        }

        fun startsWith(path: String?) = Predicate { candidate: String -> candidate.startsWith(path!!) }
    }
}
